File: C++/27.业务代表模式/BusinessDelegatePatternDemo.cpp
/**
 * @file BusinessDelegatePatternDemo.cpp
 * @author {Layne} ({shu_huanglei@163.com})
 * @brief
 * @version 0.1
 * @date 2019-12-03
 *
 * @copyright Copyright (c) 2019
 *
 */

#include "BusinessDelegate.h"
#include "Client.h"

int main(int argc, char const *argv[]) {
    std::shared_ptr<BusinessDelegate> business_delegate = std::make_shared<BusinessDelegate>();
    business_delegate->SetServiceType("EJB");

    std::shared_ptr<Client> client = std::make_shared<Client>(business_delegate);
    client->DoTask();

    business_delegate->SetServiceType("JSM");
    client->DoTask();
    return 0;
}

File: Objective-C/27.BusinessDelegatePattern/BusinessDelegatePatternDemo.m
//
//  BusinessDelegatePatternDemo.m
//  27.BusinessDelegatePattern
//
//  Created by Layne on 2020/11/21.
//

#import "BusinessDelegate.h"
#import "Client.h"

int main(int argc, const char* argv[]) {
    @autoreleasepool {
        BusinessDelegate* businessDelegate = [[BusinessDelegate alloc] init];
        [businessDelegate setServiceType:@"EJB"];

        Client* client = [[Client alloc] initWithBusinessDelegate:businessDelegate];
        [client doTask];

        [businessDelegate setServiceType:@"JMS"];
        [client doTask];
    }
    return 0;
}
